subprojects/core/src/main/java/org/gradle/internal/build/PlannedNodeGraph.java

```diff
@@ -0,0 +1,235 @@
+            this.converterRegistry = converterRegistry;
+            this.detailLevel = DetailLevel.from(converterRegistry.getConvertedNodeTypes());
+        }
+
+        public DetailLevel getDetailLevel() {
+            return detailLevel;
+        }
+
+        public void collectNodes(Collection<Node> nodes) {
+            for (Node node : nodes) {
+                ToPlannedNodeConverter converter = converterRegistry.getConverter(node);
+                if (converter != null && converter.isInSamePlan(node)) {
+                    List<? extends NodeIdentity> nodeDependencies = findNodeDependencies(node);
+                    PlannedNodeInternal plannedNode = converter.convert(node, nodeDependencies, () -> findTaskDependencies(node));
+                    plannedNodes.add(plannedNode);
+                }
+            }
+        }
+
+        public PlannedNodeGraph getGraph() {
+            return new PlannedNodeGraph(detailLevel, plannedNodes);
+        }
+
+        private List<? extends NodeIdentity> findNodeDependencies(Node node) {
+            return findDependencies(node, Node::getDependencySuccessors, it -> true);
+        }
+
+        private List<TaskIdentity> findTaskDependencies(Node node) {
+            @SuppressWarnings("unchecked")
+            List<TaskIdentity> dependencies = (List<TaskIdentity>) findDependencies(node, Node::getDependencySuccessors, TaskIdentity.class::isInstance);
+            return dependencies;
+        }
+
+        private List<? extends NodeIdentity> findDependencies(
+            Node node,
+            Function<? super Node, ? extends Collection<Node>> traverser,
+            Predicate<NodeIdentity> identityFilter
+        ) {
+            return computeDependencies(traverser, n -> getNodeIdentityOrNull(n, identityFilter), node);
+        }
+
+        private NodeIdentity getNodeIdentityOrNull(Node node, Predicate<NodeIdentity> identityFilter) {
+            ToPlannedNodeConverter converter = converterRegistry.getConverter(node);
+            if (converter == null) {
+                return null;
+            }
+
+            // Cache all identities even if the current predicate will filter them out afterwards
+            NodeIdentity nodeIdentity = nodeIdentityCache.computeIfAbsent(node, converter::getNodeIdentity);
+            return identityFilter.test(nodeIdentity) ? nodeIdentity : null;
+        }
+    }
+
+    /**
+     * Computes dependencies of a node in breadth-first order, stopping at each dependency for which identity is provided.
+     */
+    private static <T> List<NodeIdentity> computeDependencies(
+        Function<? super T, ? extends Collection<? extends T>> traverser,
+        Function<T, NodeIdentity> maybeIdentityProvider,
+        T start
+    ) {
+        List<NodeIdentity> resultDependencies = new ArrayList<>();
+
+        Queue<T> queue = new ArrayDeque<>(traverser.apply(start));
+        Set<T> seen = newSetFromMap(new IdentityHashMap<>());
```

Comment:
❓ Why not use a regular `HashSet` here? I don't see the extra value of using an "`IdentityHashSet<>`"?

Reply:
I kept the `IdentityHashSet` since it was originally used for the node traversal (`T` is `Node`).

https://github.com/gradle/gradle/blob/1429d75f592597c4f590eaaecc1e369da43a6da8/subprojects/core/src/main/java/org/gradle/internal/build/BuildOperationFiringBuildWorkPreparer.java#L164

The identity set was added here: https://github.com/gradle/gradle/commit/2cc0a9fd7e59c7876fe2fac480124f07ea31be5b

I guess it was an overprotection from some `Node`s that could potentially be equal while still being separate nodes, and thus messing up the dependency discovery. Looking at the `Node` hierarchy now, none of them override `equals`.

Given that `Node`s are non-trivial mutable objects, I would normally not use they as hash keys. However, here, it's a local state that gets thrown away and nodes are not mutated. So it should always be safe to use a regular `HashSet`.

I'll change it to the `HashSet`

Reply:
Done